### PR TITLE
JDK-8312018: Improve zero-base-optimized reservation of class space

### PR DESCRIPTION
**Pull request by tstuefe**
TL;DR This patch introduces a new reservation API to reserve memory in low address space; depending on the OS, it may use optimized placement techniques. That new API is used to optimize the placement of class space and CDS for zero-based encoding. 

A future RFE will use the same API to optimize the zero-based heap reservation and thereby consolidate a lot of coding. We also plan to use this API in other places, e.g. for Shenandoah CollectionSet reservation.

-------

With CDS off or at dump time, we currently attempt to optimize class space location by reserving in low address ranges.

We do this by examining the *java heap* end (which has been allocated at that point) and, if that had been allocated in lower address regions, attempt to allocate adjacent to it. Essentially, we piggyback on what we hope for is an optimized heap placement. If that fails, we attempt to map at HeapBaseMinAddress.

This approach has many disadvantages:

- it depends on the VM either using CompressedOops and getting a zero-based heap or the region around HeapBaseMinAddress being free. 
 
- HeapBaseMinAddress is an odd choice: it is 2G on all platforms, for reasons unknown to me, but that denies us half of the valuable low address range below 4G right away.

- We only get 1 shot. It's either one of these two addresses.

- And we only use this strategy for CDS=off or CDS=dumptime; we don't use it for the CDS-runtime-fallback-case when attaching to the primary attach point failed.

- It assumes narrow Klass encoding uses the same geometry (bit size, shift) as narrow Oops, which is not guaranteed with future developments (lilliput).

- It actually reduces the chance of getting a zero-based *java heap*. This is because when attempting to place the heap, we leave a gap for what we assume will be the later class space. That gap is `CompressedClassSpaceSize` bytes, which is often grossly over-dimensioned. A zero-based heap is more valuable than a zero-based class space. Therefore the heap should get the best chance of low-address heap reservation.

- It introduces an unnecessary dependency between heap reservation, narrow Oop encoding, and class space reservation. That makes the code base brittle.

- Getting the heap region to place class space adjacent to it is actually tricky. We lack a common get-heaprange-API because ZGC. This code misuses the CompressedOops interface. But CompressedOops is the encoding range and thus only loosely correlated to the heap range (the latter must contain the former). The fact that `CompressedOops::end()` returns the heap range end can be seen as aberration since the actual encoding range end usually far outreaches the heap range end. But as long this code relies on it returning the heap range end, we cannot fix that.

-------------

This RFE instead proposes a different approach:

- Let us have an API, `os::attempt_reserve_memory_below(address max, ...)`. This API will do its best to reserve memory with a given size and alignment below a given max address. It will, on supporting OSes, attempt to use OS-specific means to find a suitable address space hole to place the reservation in. Otherwise, it will do the typical ladder-reservation approach with an adjustable maximum number of tries.

- Let's use this API to reserve zero-base-friendly class space. Let's remove all knowledge of heap and CompressedOops. Now we are independent of what the heap does. It may or may not be located in lower address ranges. If it is, the new API will work around it and find a gap to place the class space if it is not, even better.

- Let's remove the "leave a gap for class space" logic from Heap reservation. We don't need it. It is harmful: Heap should have the best chance for zero-based - if I only can have one, I rather have a zero-based heap than a zero-based class space.

The end result will be a JVM with a much better chance to get zero-based class space *and* zero-based heap; we will have removed dependencies between heap and class space; we will have an API that can be used for similar problems (e.g. an obvious future enhancement would be to use this new reservation API for zero-based heap reservation as well, and other places could use it too, eg. Shenandoah CollectionSet reservation).

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [ ] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8312018](https://bugs.openjdk.org/browse/JDK-8312018): Improve zero-base-optimized reservation of class space (**Enhancement** - P3)

### Reviewers
 * [Roman Kennke](https://openjdk.org/census#rkennke) (@rkennke - **Reviewer**) ⚠️ Review applies to [8bb7d705](https://git.openjdk.org/jdk/pull/14867/files/8bb7d70529bdd1f945c234bbec01cca3274c28d2)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/14867/head:pull/14867` \
`$ git checkout pull/14867`

Update a local copy of the PR: \
`$ git checkout pull/14867` \
`$ git pull https://git.openjdk.org/jdk.git pull/14867/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 14867`

View PR using the GUI difftool: \
`$ git pr show -t 14867`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/14867.diff">https://git.openjdk.org/jdk/pull/14867.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/14867#issuecomment-1635627216)

### CODE REVIEW

**Comment by tstuefe** [comment]
Pinging @iklam

File: src/hotspot/share/memory/metaspace.cpp [hunk 1]
```diff
@@ -589,8 +589,31 @@ bool Metaspace::class_space_is_initialized() {
...
+
+  // First, attempt to reserve the class space such that zero-based addressing is possible.
+  // If the heap has also be allocated in low address regions, this will - with a high probability -
+  // cause the reserved range to nestle alongside the heap.
+  if (!rs.is_reserved()) {
+    // First try for zero-base zero-shift (lower 4G); failing that, try for zero-based with max shift (lower 32G)
+    constexpr uintptr_t upper_limits[2] = { UINT_MAX, KlassEncodingMetaspaceMax };
+    constexpr int num_tries = 8;
+    char* addr = nullptr;
+    for (int i = 0; i < 2 && addr == nullptr; i ++) {
```
Comment (rkennke) [inline src/hotspot/share/memory/metaspace.cpp:602]: I think it would be clearer to simply unroll this loop.

File: src/hotspot/os/linux/os_linux.cpp [hunk 2]
```diff
@@ -4226,6 +4226,90 @@ char* os::pd_attempt_reserve_memory_at(char* requested_addr, size_t bytes, bool
...
+  // attempt to map into that hole.
+
+  log_trace(os, map)("attempt_reserve_memory_below " PTR_FORMAT ", bytes: " SIZE_FORMAT_X
+                     ", alignment: " SIZE_FORMAT_X, p2i(max), bytes, alignment);
+
+// somewhat shorter
+#define ALGNUP(x) align_up(x, alignment)
+#define ALGNDWN(x) align_down(x, alignment)
+
+  FILE* f = os::fopen("/proc/self/maps", "r");
```
Comment (rkennke) [inline src/hotspot/os/linux/os_linux.cpp:4245]: How sane is it to parse /proc stuff, especially when done line-by-line? Is this not racy? I am not familiar with these things. I assume there is no sane(r) API for the purpose?

**Review by rkennke — COMMENTED** [review]
I am not familiar enough with these places of HotSpot (and the OSes, for that matter), but I have questions/comments.

**Comment by tstuefe** [comment]
> I am not familiar enough with these places of HotSpot (and the OSes, for that matter), but I have questions/comments.

Thank you Roman. I worked in your feedback; while testing I found an off-by-one, and a minor flaw with tracing.

About your procfs question, this should be safe. We only do this once, at start, and have a reasonable fallback. Note that hotspot already reads from this file for other purposes, it seems to work well.

**Review by rkennke — APPROVED** [review]
Looks good to me now, but somebody else who is more familiar with these things should review it as well. Thank you!

File: src/hotspot/share/memory/metaspace.cpp [hunk 3]
```diff
@@ -786,25 +805,7 @@ void Metaspace::global_initialize() {
...
-      if (base != nullptr) {
-        if (CompressedKlassPointers::is_valid_base(base)) {
-          rs = ReservedSpace(size, Metaspace::reserve_alignment(),
-                             os::vm_page_size(), (char*)base);
-        }
-      }
-    }
-
-    // ...failing that, reserve anywhere, but let platform do optimized placement:
+    // ...otherwise let JVM chose the best placing:
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/metaspace.cpp:808]: s/chose/choose/

File: src/hotspot/share/utilities/globalDefinitions.hpp [hunk 4]
```diff
@@ -180,6 +180,11 @@ inline intptr_t p2i(const volatile void* p) {
   return (intptr_t) p;
 }
 
+// Convert pointer to uintptr_t
+inline uintptr_t p2u(const volatile void* p) {
+  return (uintptr_t) p;
+}
+
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/globalDefinitions.hpp:187]: Seems overkill for one use.

**Comment by tstuefe** [comment]
Thanks @dholmes-ora . I worked in your feedback.

**Comment by iklam** [comment]
I think we should document the interaction with "-Xshare:dump". Maybe we should add comments that the value of `CompressedKlass::base()` is irrelevant to the dumped CDS archive when running "java -Xshared:dump", because of this code.

I.e., if CDS is enabled, we always use a non-zero based encoding.

```
narrowKlass ArchiveBuilder::get_requested_narrow_klass(Klass* k) {
  assert(DumpSharedSpaces, "sanity");
  k = get_buffered_klass(k);
  Klass* requested_k = to_requested(k);
  address narrow_klass_base = _requested_static_archive_bottom; // runtime encoding base == runtime mapping start
  const int narrow_klass_shift = ArchiveHeapWriter::precomputed_narrow_klass_shift;
  return CompressedKlassPointers::encode_not_null(requested_k, narrow_klass_base, narrow_klass_shift);
}
```

File: src/hotspot/share/runtime/os.cpp [hunk 5]
```diff
@@ -1764,13 +1764,104 @@ char* os::attempt_reserve_memory_at(char* addr, size_t bytes, bool executable) {
...
+
+char* os::get_highest_attach_address() {
+  return (char*)(
+#ifdef _LP64
+  (128 * 1024 * G)
+#else
+  SIZE_MAX
+#endif
+  - os::vm_page_size());
+}
```
Comment (iklam) [inline src/hotspot/share/runtime/os.cpp:1787]: I am not sure what "attach" means in this sense. If it's the usable address range, wouldn't it need to be OS-specific?

File: src/hotspot/share/memory/metaspace.cpp [hunk 6]
```diff
@@ -589,8 +589,28 @@ bool Metaspace::class_space_is_initialized() {
 // On error, returns an unreserved space.
 ReservedSpace Metaspace::reserve_address_space_for_compressed_classes(size_t size) {
 
+  // First, attempt to reserve the class space such that zero-based addressing is possible.
+  // If the heap has also be allocated in low address regions, this will - with a high probability -
+  // cause the reserved range to nestle alongside the heap.
+  {
+    // First try for zero-base zero-shift (lower 4G); failing that, try for zero-based with max shift (lower 32G)
+    constexpr int num_tries = 8;
```
Comment (iklam) [inline src/hotspot/share/memory/metaspace.cpp:597]: num_tries should be computed instead of hard-coded.

**Comment by tstuefe** [comment]
> I think we should document the interaction with "-Xshare:dump". Maybe we should add comments that the value of `CompressedKlass::base()` is irrelevant to the dumped CDS archive when running "java -Xshared:dump", because of this code.
> 

Okay. Maybe in a different RFE? Since its a bit tangential to what this patch does.

> I.e., if CDS is enabled, we always use a non-zero based encoding.

Not necessarily. If CDS is enabled and we don't get the preferred mapping address, we will fallback to traditional Klass range reservation and potentially go zero based. With this patch, that path is optimized too.

File: src/hotspot/share/runtime/os.cpp (hunk 5 above)
Comment (tstuefe) [reply src/hotspot/share/runtime/os.cpp:1787]: Yes, highest and lowest usable address. You are right in that its OS specific. I'll move these to the respective OS files.

File: src/hotspot/share/memory/metaspace.cpp (hunk 6 above)
Comment (tstuefe) [reply src/hotspot/share/memory/metaspace.cpp:597]: Why? In pre-existing code that does similar things, we always hardcode them implicitly (typically by attempt-mapping from A->B in hardcoded stride C). And how would I calculate it?

Comment (iklam) [reply src/hotspot/share/memory/metaspace.cpp:597]: Hmm, I am a bit unsure about the meaning of the `num_tries` parameter. It looks like Windows and Linux will scan the system's memory map and look for the first hole that's larger than `size`. However, if there are a lot of small holes at lower addresses, then doing 8 tries won't find a large enough block, even though such a block may exist below `unscaled_max`.

The default implementation will try to find a free block with fixed steps. In this case, it seems like 

```
num_tries = (unscaled_max + size - 1) / size;
```

would be more appropriate. Otherwise if `size` changes in the future (to be smaller), again you won't be able to find an appropriate block, even if one exists.

So I am not sure if the caller passing in an arbitrary `num_tries` parameter is a good idea. Maybe the API needs to be redesigned.

**Comment by tstuefe** [comment]
@iklam @rkennke @dholmes-ora Thanks for your feedback. I'll close this PR in favor of a fresh one, since I did some considerable changes to the API and I don't want to flood your mailboxes with skara spam.
